File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
  const solution = [];
  function backTrack(index) {
    if (index === nums.length - 1) return solution.push(nums.slice());
    
    for (let i = index; i < nums.length; i++) {
      [nums[i], nums[index]] = [nums[index], nums[i]];
      backTrack(index + 1);
      [nums[i], nums[index]] = [nums[index], nums[i]];
    }
  }
  
  backTrack(0);
  return solution;
};

File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if len(nums) == 0:
            return [[]]
        
        first = nums[0]
        perms = self.permute(nums[1:])
        all_perms = []
        
        for perm in perms:
            for i in range(len(perm) + 1):
                all_perms.append(perm[:i] + [first] + perm[i:])
        
        return all_perms